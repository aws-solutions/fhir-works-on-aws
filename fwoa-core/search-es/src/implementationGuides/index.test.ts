/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { SearchImplementationGuides } from './index';

const { compile } = SearchImplementationGuides;

describe('compile', () => {
  test(`simple path - Patient.communication.language`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/Patient-language',
        name: 'language',
        code: 'language',
        type: 'token',
        description: 'Language code (irrespective of use value)',
        base: ['Patient'],
        expression: 'Patient.communication.language',
        xpath: 'f:Patient/f:communication/f:language'
      }
    ]);

    await expect(compiled).resolves.toMatchSnapshot();
  });

  test(`simple where - Library.relatedArtifact.where(type='predecessor').resource`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/Library-predecessor',
        name: 'predecessor',
        code: 'predecessor',
        description: 'What resource is being referenced',
        base: ['Library'],
        type: 'reference',
        expression: "Library.relatedArtifact.where(type='predecessor').resource",
        xpath: "f:Library/f:relatedArtifact[f:type/@value='predecessor']/f:resource",
        target: ['Account', 'ActivityDefinition']
      }
    ]);
    await expect(compiled).resolves.toMatchSnapshot();
  });

  test(`where with resolve() is - Person.link.target.where(resolve() is RelatedPerson)`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/Person-relatedperson',
        name: 'relatedperson',
        code: 'relatedperson',
        type: 'reference',
        description: 'The Person links to this RelatedPerson',
        base: ['Person'],
        expression: 'Person.link.target.where(resolve() is RelatedPerson)',
        xpath: 'f:Person/f:link/f:target',
        target: ['RelatedPerson']
      }
    ]);
    await expect(compiled).resolves.toMatchSnapshot();
  });
  test(`as - (ConceptMap.source as uri)`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/ConceptMap-source-uri',
        name: 'source-uri',
        code: 'source-uri',
        type: 'reference',
        description: 'The source value set that contains the concepts that are being mapped',
        base: ['ConceptMap'],
        expression: '(ConceptMap.source as uri)',
        xpath: 'f:ConceptMap/f:sourceUri',
        target: ['ValueSet']
      }
    ]);
    await expect(compiled).resolves.toMatchSnapshot();
  });
  test(`OR operator - CapabilityStatement.title | CodeSystem.title | ConceptMap.title | ImplementationGuide.title | MessageDefinition.title | OperationDefinition.title | StructureDefinition.title | StructureMap.title | TerminologyCapabilities.title | ValueSet.title`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/conformance-title',
        name: 'title',
        code: 'title',
        type: 'string',
        description: 'Multiple Resources...',
        base: [
          'CapabilityStatement',
          'CodeSystem',
          'ConceptMap',
          'ImplementationGuide',
          'MessageDefinition',
          'OperationDefinition',
          'StructureDefinition',
          'StructureMap',
          'TerminologyCapabilities',
          'ValueSet'
        ],
        expression:
          'CapabilityStatement.title | CodeSystem.title | ConceptMap.title | ImplementationGuide.title | MessageDefinition.title | OperationDefinition.title | StructureDefinition.title | StructureMap.title | TerminologyCapabilities.title | ValueSet.title',
        xpath:
          'f:CapabilityStatement/f:title | f:CodeSystem/f:title | f:ConceptMap/f:title | f:ImplementationGuide/f:title | f:MessageDefinition/f:title | f:OperationDefinition/f:title | f:StructureDefinition/f:title | f:StructureMap/f:title | f:TerminologyCapabilities/f:title | f:ValueSet/f:title'
      }
    ]);
    await expect(compiled).resolves.toMatchSnapshot();
  });
  test(`OR operator with same base resource - (ExampleScenario.useContext.value as Quantity) | (ExampleScenario.useContext.value as Range)`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/ExampleScenario-context-quantity',
        name: 'context-quantity',
        code: 'context-quantity',
        type: 'quantity',
        description: 'A quantity- or range-valued use context assigned to the example scenario',
        base: ['ExampleScenario'],
        expression:
          '(ExampleScenario.useContext.value as Quantity) | (ExampleScenario.useContext.value as Range)',
        xpath: 'f:ExampleScenario/f:useContext/f:valueQuantity | f:ExampleScenario/f:useContext/f:valueRange'
      }
    ]);
    await expect(compiled).resolves.toMatchSnapshot();
  });

  test(`simple where url value - Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.value.code`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/test',
        name: 'test',
        code: 'test',
        type: 'token',
        description: 'test',
        base: ['Patient'],
        expression:
          "Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.value.code",
        xpath:
          "f:Patient/f:extension[@url='http://hl7.org/fhir/us/core/StructureDefinition/us-core-race']/f:extension/f:valueCoding/f:code/@value"
      }
    ]);
    await expect(compiled).resolves.toMatchSnapshot();
  });

  test(`.extension() function - DocumentReference.extension('http://example.org/fhir/StructureDefinition/participation-agreement')`, async () => {
    // example extension from https://www.hl7.org/fhir/R4/searchparameter-example-extension.json.html
    const extensionSearchParam = {
      resourceType: 'SearchParameter',
      id: 'example-extension',
      url: 'http://hl7.org/fhir/SearchParameter/example-extension',
      name: 'Example Search Parameter on an extension',
      description: 'Search by url for a participation agreement, which is stored in a DocumentReference',
      code: 'part-agree',
      base: ['DocumentReference'],
      type: 'reference',
      expression:
        "DocumentReference.extension('http://example.org/fhir/StructureDefinition/participation-agreement')",
      xpath:
        "f:DocumentReference/f:extension[@url='http://example.org/fhir/StructureDefinition/participation-agreement']",
      target: ['DocumentReference']
    };
    const compiled = compile([extensionSearchParam]);
    await expect(compiled).resolves.toMatchSnapshot();

    // .extension('x') is syntactic sugar for extension.where(url = 'x')
    await expect(compiled).resolves.toEqual(
      await compile([
        {
          ...extensionSearchParam,
          expression: `DocumentReference.extension.where(url = 'http://example.org/fhir/StructureDefinition/participation-agreement')`
        }
      ])
    );
  });

  test(`xpath explicitly expands choice of data types and fhirPath does not`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/Consent-source-reference',
        name: 'source-reference',
        code: 'source-reference',
        type: 'reference',
        description: 'test',
        base: ['Consent'],
        expression: 'Consent.source',
        xpath: 'f:Consent/f:sourceAttachment | f:Consent/f:sourceReference',
        target: ['ValueSet']
      }
    ]);
    await expect(compiled).resolves.toMatchSnapshot();
  });

  test(`fhirPath with broader conditions than xPath`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/Device-device-name',
        name: 'device-name',
        code: 'device-name',
        type: 'string',
        description: 'test',
        base: ['Device'],
        expression: 'Device.deviceName.name | Device.type.coding.display | Device.type.text',
        xpath: 'f:Device/f:deviceName'
      }
    ]);
    await expect(compiled).resolves.toMatchSnapshot();
  });

  test(`expression with extra whitespaces`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/test',
        name: 'test',
        code: 'test',
        type: 'token',
        description: 'test',
        base: ['Patient'],
        expression: "Patient.x.where(field         =         'value')",
        xpath: "f:Patient/f:x[f:field  =  'value']"
      }
    ]);
    await expect(compiled).resolves.toMatchSnapshot();
  });

  test(`Invalid input`, async () => {
    const compiled = compile([
      {
        foo: 'bar'
      }
    ]);
    await expect(compiled).rejects.toThrowError();
  });

  test(`unparsable FHIRPath expression`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/test',
        name: 'test',
        code: 'test',
        type: 'token',
        description: 'test',
        base: ['Patient'],
        expression: 'some random FHIRPath expression that cannot be parsed',
        xpath: 'some random FHIRPath expression that cannot be parsed'
      }
    ]);
    await expect(compiled).rejects.toThrowError();
  });
  test(`search param of type reference with no target`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/ConceptMap-source-uri',
        name: 'source-uri',
        code: 'source-uri',
        type: 'reference',
        description: 'The source value set that contains the concepts that are being mapped',
        base: ['ConceptMap'],
        expression: '(ConceptMap.source as uri)',
        xpath: 'f:ConceptMap/f:sourceUri'
      }
    ]);
    await expect(compiled).rejects.toThrowError();
  });

  test(`search param of type reference with empty target`, async () => {
    const compiled = compile([
      {
        resourceType: 'SearchParameter',
        url: 'http://hl7.org/fhir/SearchParameter/ConceptMap-source-uri',
        name: 'source-uri',
        code: 'source-uri',
        type: 'reference',
        description: 'The source value set that contains the concepts that are being mapped',
        base: ['ConceptMap'],
        expression: '(ConceptMap.source as uri)',
        xpath: 'f:ConceptMap/f:sourceUri',
        target: []
      }
    ]);
    await expect(compiled).rejects.toThrowError();
  });
});
