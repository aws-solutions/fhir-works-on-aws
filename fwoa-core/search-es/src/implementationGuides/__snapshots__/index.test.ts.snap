// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile .extension() function - DocumentReference.extension('http://example.org/fhir/StructureDefinition/participation-agreement') 1`] = `
Array [
  Object {
    "base": "DocumentReference",
    "compiled": Array [
      Object {
        "condition": Array [
          "extension.url",
          "=",
          "http://example.org/fhir/StructureDefinition/participation-agreement",
        ],
        "path": "extension",
        "resourceType": "DocumentReference",
      },
    ],
    "description": "Search by url for a participation agreement, which is stored in a DocumentReference",
    "name": "part-agree",
    "target": Array [
      "DocumentReference",
    ],
    "type": "reference",
    "url": "http://hl7.org/fhir/SearchParameter/example-extension",
  },
]
`;

exports[`compile OR operator - CapabilityStatement.title | CodeSystem.title | ConceptMap.title | ImplementationGuide.title | MessageDefinition.title | OperationDefinition.title | StructureDefinition.title | StructureMap.title | TerminologyCapabilities.title | ValueSet.title 1`] = `
Array [
  Object {
    "base": "CapabilityStatement",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "CapabilityStatement",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
  Object {
    "base": "CodeSystem",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "CodeSystem",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
  Object {
    "base": "ConceptMap",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "ConceptMap",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
  Object {
    "base": "ImplementationGuide",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "ImplementationGuide",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
  Object {
    "base": "MessageDefinition",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "MessageDefinition",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
  Object {
    "base": "OperationDefinition",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "OperationDefinition",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
  Object {
    "base": "StructureDefinition",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "StructureDefinition",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
  Object {
    "base": "StructureMap",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "StructureMap",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
  Object {
    "base": "TerminologyCapabilities",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "TerminologyCapabilities",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
  Object {
    "base": "ValueSet",
    "compiled": Array [
      Object {
        "path": "title",
        "resourceType": "ValueSet",
      },
    ],
    "description": "Multiple Resources...",
    "name": "title",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/conformance-title",
  },
]
`;

exports[`compile OR operator with same base resource - (ExampleScenario.useContext.value as Quantity) | (ExampleScenario.useContext.value as Range) 1`] = `
Array [
  Object {
    "base": "ExampleScenario",
    "compiled": Array [
      Object {
        "path": "useContext.valueQuantity",
        "resourceType": "ExampleScenario",
      },
      Object {
        "path": "useContext.valueRange",
        "resourceType": "ExampleScenario",
      },
    ],
    "description": "A quantity- or range-valued use context assigned to the example scenario",
    "name": "context-quantity",
    "target": undefined,
    "type": "quantity",
    "url": "http://hl7.org/fhir/SearchParameter/ExampleScenario-context-quantity",
  },
]
`;

exports[`compile as - (ConceptMap.source as uri) 1`] = `
Array [
  Object {
    "base": "ConceptMap",
    "compiled": Array [
      Object {
        "path": "sourceUri",
        "resourceType": "ConceptMap",
      },
    ],
    "description": "The source value set that contains the concepts that are being mapped",
    "name": "source-uri",
    "target": Array [
      "ValueSet",
    ],
    "type": "reference",
    "url": "http://hl7.org/fhir/SearchParameter/ConceptMap-source-uri",
  },
]
`;

exports[`compile expression with extra whitespaces 1`] = `
Array [
  Object {
    "base": "Patient",
    "compiled": Array [
      Object {
        "condition": Array [
          "x.field",
          "=",
          "value",
        ],
        "path": "x",
        "resourceType": "Patient",
      },
    ],
    "description": "test",
    "name": "test",
    "target": undefined,
    "type": "token",
    "url": "http://hl7.org/fhir/SearchParameter/test",
  },
]
`;

exports[`compile fhirPath with broader conditions than xPath 1`] = `
Array [
  Object {
    "base": "Device",
    "compiled": Array [
      Object {
        "path": "deviceName.name",
        "resourceType": "Device",
      },
      Object {
        "path": "type.coding.display",
        "resourceType": "Device",
      },
      Object {
        "path": "type.text",
        "resourceType": "Device",
      },
      Object {
        "path": "deviceName",
        "resourceType": "Device",
      },
    ],
    "description": "test",
    "name": "device-name",
    "target": undefined,
    "type": "string",
    "url": "http://hl7.org/fhir/SearchParameter/Device-device-name",
  },
]
`;

exports[`compile simple path - Patient.communication.language 1`] = `
Array [
  Object {
    "base": "Patient",
    "compiled": Array [
      Object {
        "path": "communication.language",
        "resourceType": "Patient",
      },
    ],
    "description": "Language code (irrespective of use value)",
    "name": "language",
    "target": undefined,
    "type": "token",
    "url": "http://hl7.org/fhir/SearchParameter/Patient-language",
  },
]
`;

exports[`compile simple where - Library.relatedArtifact.where(type='predecessor').resource 1`] = `
Array [
  Object {
    "base": "Library",
    "compiled": Array [
      Object {
        "condition": Array [
          "relatedArtifact.type",
          "=",
          "predecessor",
        ],
        "path": "relatedArtifact.resource",
        "resourceType": "Library",
      },
    ],
    "description": "What resource is being referenced",
    "name": "predecessor",
    "target": Array [
      "Account",
      "ActivityDefinition",
    ],
    "type": "reference",
    "url": "http://hl7.org/fhir/SearchParameter/Library-predecessor",
  },
]
`;

exports[`compile simple where url value - Patient.extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.value.code 1`] = `
Array [
  Object {
    "base": "Patient",
    "compiled": Array [
      Object {
        "condition": Array [
          "extension.url",
          "=",
          "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race",
        ],
        "path": "extension.extension.value.code",
        "resourceType": "Patient",
      },
      Object {
        "condition": Array [
          "extension.url",
          "=",
          "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race",
        ],
        "path": "extension.extension.valueCoding.code",
        "resourceType": "Patient",
      },
    ],
    "description": "test",
    "name": "test",
    "target": undefined,
    "type": "token",
    "url": "http://hl7.org/fhir/SearchParameter/test",
  },
]
`;

exports[`compile where with resolve() is - Person.link.target.where(resolve() is RelatedPerson) 1`] = `
Array [
  Object {
    "base": "Person",
    "compiled": Array [
      Object {
        "condition": Array [
          "link.target",
          "resolve",
          "RelatedPerson",
        ],
        "path": "link.target",
        "resourceType": "Person",
      },
    ],
    "description": "The Person links to this RelatedPerson",
    "name": "relatedperson",
    "target": Array [
      "RelatedPerson",
    ],
    "type": "reference",
    "url": "http://hl7.org/fhir/SearchParameter/Person-relatedperson",
  },
]
`;

exports[`compile xpath explicitly expands choice of data types and fhirPath does not 1`] = `
Array [
  Object {
    "base": "Consent",
    "compiled": Array [
      Object {
        "path": "source",
        "resourceType": "Consent",
      },
      Object {
        "path": "sourceAttachment",
        "resourceType": "Consent",
      },
      Object {
        "path": "sourceReference",
        "resourceType": "Consent",
      },
    ],
    "description": "test",
    "name": "source-reference",
    "target": Array [
      "ValueSet",
    ],
    "type": "reference",
    "url": "http://hl7.org/fhir/SearchParameter/Consent-source-reference",
  },
]
`;
